src/utility/parser.cpp

```diff
@@ -0,0 +1,278 @@
+#include <BSMPT/config.h>
+#include <BSMPT/utility/Logger.h>
+#include <BSMPT/utility/parser.h>
+#include <BSMPT/utility/utility.h>
+#include <algorithm>
+#include <sstream>
+
+#ifdef nlohmann_json_FOUND
+#include <fstream>
+#include <nlohmann/json.hpp>
+#endif
+
+namespace BSMPT
+{
+
+parserException::parserException(const std::string &msg) : message{msg}
+{
+}
+
+const char *parserException::what() const noexcept
+{
+  return message.c_str();
+}
+
+parser::parser()
+{
+  add_argument("logginglevel::disabled", "Disable the Logger.", false);
+  add_argument("logginglevel::debug",
+               "Turn on/off (true/false) the debug output for the logger.",
+               false);
+  add_argument("logginglevel::ewbgdetailed",
+               "Turn on/off (true/false) the output for the EWBG calculation.",
+               false);
+  add_argument(
+      "logginglevel::progdetailed",
+      "Turn on/off (true/false) the additional output for the main program.",
+      false);
+  add_argument(
+      "logginglevel::minimizerdetailed",
+      "Turn on/off (true/false) the additional output of the minimizers.",
+      false);
+  add_argument("logginglevel::default",
+               "Turn on/off (true/false) the default output.",
+               false);
+
+  add_argument("json",
+               "Use a json file to define the input instead of additional cli "
+               "parameters.",
+               false);
+}
+
+void parser::enable_minimizer_options()
+{
+  add_argument("useGSL",
+               "Use the GSL library to minimize the effective potential.",
+               false);
+  add_argument("useCMAES",
+               "Use the CMAES library to minimize the effective potential",
+               false);
+  add_argument("useNLopt",
+               "Use the NLopt library to minimize the effective potential",
+               false);
+  add_argument("useMultithreading",
+               "Enables/Disables multi threading for the minimizers",
+               false);
+}
+
+void parser::add_argument(const std::string &argument,
+                          const std::string &description,
+                          bool required)
+{
+  auto arg = to_lower(argument);
+  Options options;
+  options.argument    = argument;
+  options.description = description;
+  options.value       = std::nullopt;
+  if (required)
+  {
+    mRequiredArguments.emplace(arg, options);
+  }
+  else
+  {
+    mOptionalArguments.emplace(arg, options);
+  }
+}
+
+void parser::add_input(const std::vector<std::string> &input)
+{
+  if (input.size() == 1)
+  {
+    std::string arg{input.at(0)};
+    auto [argument, value] = get_key_value(arg);
+    if (argument == "help")
```

Comment:
Comment on the `--help` mode: Because in all executables, we call `check_required_parameters()` again after `add_argument()`, calling the executable with `--help` results in the usual help output twice, because of `check_required_parameters()` calling `print_help()` again for the case of missing parameters. Is this a wanted behaviour? In my view, displaying the help once is sufficient and would make the output more clear.

Reply:
Fixed it